protocol/swap/manager.go

```diff
@@ -1,168 +1,156 @@
 }
 
 type manager struct {
+	db Database
 	sync.RWMutex
 	ongoing map[types.Hash]*Info
 	past    map[types.Hash]*Info
 }
 
-// NewManager ...
-func NewManager() Manager {
+// NewManager returns a new Manager that uses the given database.
+// It loads all ongoing swaps into memory on construction.
+// Completed swaps are not loaded into memory.
+func NewManager(db Database) (*manager, error) {
+	ongoing := make(map[types.Hash]*Info)
+
+	stored, err := db.GetAllSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range stored {
+		if !s.Status.IsOngoing() {
+			continue
+		}
+
+		ongoing[s.ID] = s
+	}
+
 	return &manager{
-		ongoing: make(map[types.Hash]*Info),
+		db:      db,
+		ongoing: ongoing,
 		past:    make(map[types.Hash]*Info),
-	}
+	}, nil
 }
 
 // AddSwap adds the given swap *Info to the Manager.
 func (m *manager) AddSwap(info *Info) error {
 	m.Lock()
 	defer m.Unlock()
 
-	switch info.status.IsOngoing() {
+	switch info.Status.IsOngoing() {
 	case true:
-		m.ongoing[info.id] = info
+		m.ongoing[info.ID] = info
 	default:
-		m.past[info.id] = info
+		m.past[info.ID] = info
 	}
 
-	return nil
+	return m.db.PutSwap(info)
 }
 
 // GetPastIDs returns all past swap IDs.
-func (m *manager) GetPastIDs() []types.Hash {
+func (m *manager) GetPastIDs() ([]types.Hash, error) {
 	m.RLock()
 	defer m.RUnlock()
-	ids := make([]types.Hash, len(m.past))
-	i := 0
+	ids := make(map[types.Hash]struct{})
 	for id := range m.past {
-		ids[i] = id
+		ids[id] = struct{}{}
+	}
+
+	// TODO: do we want to cache all past swaps since we're already fetching them?
+	stored, err := m.db.GetAllSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range stored {
+		if s.Status.IsOngoing() {
+			continue
+		}
+
+		ids[s.ID] = struct{}{}
+	}
+
+	idArr := make([]types.Hash, len(ids))
+	i := 0
+	for id := range ids {
+		idArr[i] = id
 		i++
 	}
-	return ids
+
+	return idArr, nil
 }
 
 // GetPastSwap returns a swap's *Info given its ID.
-func (m *manager) GetPastSwap(id types.Hash) *Info {
+func (m *manager) GetPastSwap(id types.Hash) (*Info, error) {
 	m.RLock()
 	defer m.RUnlock()
-	return m.past[id]
+	s, has := m.past[id]
+	if has {
+		return s, nil
+	}
+
+	// TODO: do we want to cache this swap?
+	return m.getSwapFromDB(id)
 }
 
 // GetOngoingSwap returns the ongoing swap's *Info, if there is one.
-func (m *manager) GetOngoingSwap(id types.Hash) *Info {
+func (m *manager) GetOngoingSwap(id types.Hash) (*Info, error) {
 	m.RLock()
 	defer m.RUnlock()
-	return m.ongoing[id]
+	s, has := m.ongoing[id]
+	if !has {
+		return nil, errNoSwapWithID
+	}
+
+	return s, nil
 }
 
 // CompleteOngoingSwap marks the current ongoing swap as completed.
-func (m *manager) CompleteOngoingSwap(id types.Hash) {
+func (m *manager) CompleteOngoingSwap(id types.Hash) error {
 	m.Lock()
 	defer m.Unlock()
 	s, has := m.ongoing[id]
 	if !has {
-		return
+		return nil
 	}
 
 	m.past[id] = s
 	delete(m.ongoing, id)
+
+	// re-write to db, as status has changed
```

Comment:
I'm trying to think of how to make this more clear, but no great ideas are coming to my mind. It's a little confusing that this function is "marking" the "ongoing swap" as "completed" and updating the DB, but the status must be getting changed somewhere else before we are called.

Reply:
right, the main solutions I can think of are separating ongoing and completed swaps within the db into two separate "tables", or writing the status to a key separate than the `swap.Info` is written to